src/transformers/trainer_utils.py

```diff
@@ -510,14 +528,24 @@ def stop(self, stage):
 
         # gpu
         if self.torch is not None:
-            self.gpu_mem_used_now = self.torch.cuda.memory_allocated()
-            self.gpu_mem_used_peak = self.torch.cuda.max_memory_allocated()
-            self.gpu[self.cur_stage] = {
-                "begin": self.gpu_mem_used_at_start,
-                "end": self.gpu_mem_used_now,
-                "alloc": (self.gpu_mem_used_now - self.gpu_mem_used_at_start),
-                "peaked": max(0, self.gpu_mem_used_peak - self.gpu_mem_used_now),
-            }
+            if torch.cuda.is_available():
+                self.gpu_mem_used_now = self.torch.cuda.memory_allocated()
+                self.gpu_mem_used_peak = self.torch.cuda.max_memory_allocated()
+                self.gpu[self.cur_stage] = {
+                    "begin": self.gpu_mem_used_at_start,
+                    "end": self.gpu_mem_used_now,
+                    "alloc": (self.gpu_mem_used_now - self.gpu_mem_used_at_start),
+                    "peaked": max(0, self.gpu_mem_used_peak - self.gpu_mem_used_now),
+                }
+            elif is_torch_xpu_available():
+                self.gpu_mem_used_now = self.torch.xpu.memory_allocated()
+                self.gpu_mem_used_peak = self.torch.xpu.max_memory_allocated()
+                self.gpu[self.cur_stage] = {
+                    "begin": self.gpu_mem_used_at_start,
+                    "end": self.gpu_mem_used_now,
+                    "alloc": (self.gpu_mem_used_now - self.gpu_mem_used_at_start),
+                    "peaked": max(0, self.gpu_mem_used_peak - self.gpu_mem_used_now),
+                }
```

Comment:
We can reduce some of this repeated code 

```suggestion
            if torch.cuda.is_available():
                self.gpu_mem_used_now = self.torch.cuda.memory_allocated()
                self.gpu_mem_used_peak = self.torch.cuda.max_memory_allocated()
            elif is_torch_xpu_available():
                self.gpu_mem_used_now = self.torch.xpu.memory_allocated()
                self.gpu_mem_used_peak = self.torch.xpu.max_memory_allocated()
            else:
                raise ValueError("No available GPU device found!")
                
            self.gpu[self.cur_stage] = {
                "begin": self.gpu_mem_used_at_start,
                "end": self.gpu_mem_used_now,
                "alloc": (self.gpu_mem_used_now - self.gpu_mem_used_at_start),
                "peaked": max(0, self.gpu_mem_used_peak - self.gpu_mem_used_now),
            }
```

@muellerzr  - is it possible for `self.torch` to be set but `torch.cuda.is_available()` to be `False` ?